src/Aspire.Hosting/Dashboard/DashboardServiceHost.cs

```diff
@@ -38,76 +54,134 @@ internal sealed class DashboardServiceHost : IHostedService
+            {
+                if (!uri.IsLoopback)
                 {
-                    kestrelOptions.Listen(IPAddress.Parse(uri.Host), uri.Port, ConfigureListen);
+                    throw new ArgumentException($"{ResourceServiceUrlVariableName} must contain a local loopback address.");
                 }
 
-                void ConfigureListen(ListenOptions options)
+                scheme = uri.Scheme;
+
+                kestrelOptions.ListenLocalhost(uri.Port, ConfigureListen);
+
+            }
+            else
+            {
+                throw new ArgumentException($"Multiple URIs are not supported in the {ResourceServiceUrlVariableName} environment variable.");
+            }
+
+            void ConfigureListen(ListenOptions options)
+            {
+                // Force HTTP/2 for gRPC, so that it works over non-TLS connections
+                // which cannot negotiate between HTTP/1.1 and HTTP/2.
+                options.Protocols = HttpProtocols.Http2;
+
+                if (string.Equals(scheme, "https", StringComparison.Ordinal))
                 {
-                    // Force HTTP/2 for gRPC, so that it works over non-TLS connections
-                    // which cannot negotiate between HTTP/1.1 and HTTP/2.
-                    options.Protocols = HttpProtocols.Http2;
-
-                    if (string.Equals(uri.Scheme, "https", StringComparison.Ordinal))
-                    {
-                        options.UseHttps();
-                    }
+                    options.UseHttps();
                 }
             }
         }
     }
 
+    /// <summary>
+    /// Gets the URI upon which the resource service is listening.
+    /// </summary>
+    /// <remarks>
+    /// Intended to be used by the app model when launching the dashboard process, populating its
+    /// <c>DOTNET_RESOURCE_SERVICE_ENDPOINT_URL</c> environment variable with a single URI.
+    /// </remarks>
+    public async Task<string> GetResourceServiceUriAsync(CancellationToken cancellationToken = default)
+    {
+        var stopwatch = ValueStopwatch.StartNew();
+
+        var uri = await _resourceServiceUri.Task.WaitAsync(cancellationToken).ConfigureAwait(false);
+
+        var elapsed = stopwatch.GetElapsedTime();
+
+        if (elapsed > TimeSpan.FromSeconds(2))
+        {
+            _logger.LogWarning("Unexpectedly long wait for resource service URI ({Elapsed}).", elapsed);
+        }
+
+        return uri;
+    }
+
     async Task IHostedService.StartAsync(CancellationToken cancellationToken)
     {
         if (_app is not null)
         {
             await _app.StartAsync(cancellationToken).ConfigureAwait(false);
+
+            var addressFeature = _app.Services.GetService<IServer>()?.Features.Get<IServerAddressesFeature>();
+
+            if (addressFeature is null)
+            {
+                _resourceServiceUri.SetException(new InvalidOperationException("Could not obtain IServerAddressesFeature. Dashboard URIs are not available."));
```

Comment:
We modelled multiple values here previously, then changed to singular. This needs updating:

```suggestion
                _resourceServiceUri.SetException(new InvalidOperationException("Could not obtain IServerAddressesFeature. Resource service URI is not available."));
```